src/Parsing/Parser.hs

```diff
@@ -140,57 +141,65 @@ pythonParserPrecise = UnmarshalParser PrecisePython.tree_sitter_python
+goParser' :: c Go.Term => (Language, SomeParser c Loc)
 goParser' = (Go, SomeParser goParser)
 
-javaParser' :: c PreciseJava.Term => (Language, SomeParser c Loc)
+javaParser' :: c Java.Term => (Language, SomeParser c Loc)
 javaParser' = (Java, SomeParser javaParserPrecise)
 
-javascriptParser' :: c (Term (Sum TSX.Syntax)) => (Language, SomeParser c Loc)
+javascriptParser' :: c TSX.Term => (Language, SomeParser c Loc)
 javascriptParser' = (JavaScript, SomeParser tsxParser)
 
-jsonParserPrecise' :: c PreciseJSON.Term => (Language, SomeParser c Loc)
+jsonParserPrecise' :: c JSON.Term => (Language, SomeParser c Loc)
 jsonParserPrecise' = (JSON, SomeParser jsonParserPrecise)
 
-jsxParser' :: c (Term (Sum TSX.Syntax)) => (Language, SomeParser c Loc)
+jsxParser' :: c TSX.Term => (Language, SomeParser c Loc)
 jsxParser' = (JSX, SomeParser tsxParser)
 
-markdownParser' :: c (Term (Sum Markdown.Syntax)) => (Language, SomeParser c Loc)
+markdownParser' :: c Markdown.Term => (Language, SomeParser c Loc)
 markdownParser' = (Markdown, SomeParser markdownParser)
 
-phpParser' :: c (Term (Sum PHP.Syntax)) => (Language, SomeParser c Loc)
+phpParser' :: c PHP.Term => (Language, SomeParser c Loc)
 phpParser' = (PHP, SomeParser phpParser)
 
-pythonParserALaCarte' :: c (Term (Sum Python.Syntax)) => (Language, SomeParser c Loc)
+pythonParserALaCarte' :: c PythonALaCarte.Term => (Language, SomeParser c Loc)
 pythonParserALaCarte' = (Python, SomeParser pythonParser)
 
-pythonParserPrecise' :: c PrecisePython.Term => (Language, SomeParser c Loc)
+pythonParserPrecise' :: c PythonPrecise.Term => (Language, SomeParser c Loc)
 pythonParserPrecise' = (Python, SomeParser pythonParserPrecise)
 
-pythonParser' :: (c (Term (Sum Python.Syntax)), c PrecisePython.Term) => PerLanguageModes -> (Language, SomeParser c Loc)
+pythonParser' :: (c PythonALaCarte.Term, c PythonPrecise.Term) => PerLanguageModes -> (Language, SomeParser c Loc)
 pythonParser' modes = case pythonMode modes of
   ALaCarte -> (Python, SomeParser pythonParser)
   Precise  -> (Python, SomeParser pythonParserPrecise)
 
-rubyParser' :: c (Term (Sum Ruby.Syntax)) => (Language, SomeParser c Loc)
+rubyParser' :: c Ruby.Term => (Language, SomeParser c Loc)
 rubyParser' = (Ruby, SomeParser rubyParser)
 
-tsxParser' :: c (Term (Sum TSX.Syntax)) => (Language, SomeParser c Loc)
+tsxParser' :: c TSX.Term => (Language, SomeParser c Loc)
 tsxParser' = (TSX, SomeParser tsxParser)
 
-typescriptParser' :: c (Term (Sum TypeScript.Syntax)) => (Language, SomeParser c Loc)
+typescriptParser' :: c TypeScript.Term => (Language, SomeParser c Loc)
 typescriptParser' = (TypeScript, SomeParser typescriptParser)
 
 
+-- | A type family selecting the language mode for a given term type.
+type family TermMode term where
+  TermMode Java.Term          = 'Precise
+  TermMode JSON.Term          = 'Precise
+  TermMode PythonPrecise.Term = 'Precise
+  TermMode _                  = 'ALaCarte
```

Comment:
We can use this to select between different strategies for implementing classes using the advanced overlap trick.

Reply:
Loving it.